Don't dissoc `:id` in `before-update`

johnswanson opened the pull request:
There seems to be a [nasty footgun](https://github.com/camsaul/toucan2/pull/196) in toucan2 here.

Say you're executing an `update!` command on a set of IDs, e.g.:

```
(t2/update! :model/Card :id [:in 1 2] {:view_count 1})
```

This works if:

- both cards 1 and 2 have `view_count=1` already

- both cards 1 and 2 have `view_count!=1`

However.

If one of the two cards has `view_count=1` and another has a different view_count, then (if the `before-update` method doesn't have the primary key attached) Toucan emits a call to update *every card in the database*, without a where clause at all.

johnswanson committed:
Don't dissoc `:id` in `before-update`

There seems to be a nasty footgun in toucan2 here.

Say you're executing an `update!` command on a set of IDs, e.g.:

```
(t2/update! :model/Card :id [:in 1 2] {:view_count 1})
```

This works if:

- both cards 1 and 2 have `view_count=1` already

- both cards 1 and 2 have `view_count!=1`

However.

If one of the two cards has `view_count=1` and another has a different
view_count, then (if the `before-update` method doesn't have the primary
key attached) Toucan emits a call to update *every card in the
database*, without a where clause at all.

noahmoss commented:
@johnswanson Do you think it'd be worth adding a test around the `t2/update!` behavior you were seeing, specifically for cards? Find a case that repros it consistently when the `dissoc` call is there, so that we don't accidentally reintroduce it?

johnswanson committed:
Add a test

johnswanson committed:
Fix the test

Not sure why - on MariaDB and MySQL we seem to return `4` instead of `2`
here, but we are only changing the two rows as far as I can tell. Fixed
the test to just assert the important thing: no other rows
were *actually* changed.
